src/Agent.Worker/Build/GitSourceProvider.cs

```diff
@@ -1,11 +1,758 @@
+            context.Debug("Remove injected credential from git remote fetch url.");
+            Int32 exitCode_seturl = await GitRemoteSetUrl(context, repositoryPath, remoteName, repositoryUrl.AbsoluteUri);
+
+            context.Debug("Remove injected credential from git remote push url.");
+            Int32 exitCode_setpushurl = await GitRemoteSetPushUrl(context, repositoryPath, remoteName, repositoryUrl.AbsoluteUri);
+
+            if (exitCode_seturl != 0 || exitCode_setpushurl != 0)
+            {
+                // if unable to use git.exe set fetch url back, modify git config file on disk. make sure we don't left credential.
+                context.Debug("Unable to use git.exe remove injected credential from git remote fetch url, modify git config file on disk to remove injected credential.");
+                String gitConfig = Path.Combine(repositoryPath, ".git/config");
+                if (File.Exists(gitConfig))
+                {
+                    // TODO: async read/write ?
+                    String gitConfigContent = File.ReadAllText(Path.Combine(repositoryPath, ".git", "config"));
+                    Uri urlWithCred;
+                    if (_credentialUrlCache.TryGetValue(repositoryUrl.AbsoluteUri, out urlWithCred))
+                    {
+                        gitConfigContent = gitConfigContent.Replace(urlWithCred.AbsoluteUri, repositoryUrl.AbsoluteUri);
+                        File.WriteAllText(gitConfig, gitConfigContent);
+                    }
+                }
+            }
+        }
+
+        private String GetCommandOption(String command, Version version)
+        {
+            if (String.IsNullOrEmpty(command))
+            {
+                throw new ArgumentNullException("command");
+            }
+
+            if (version < _minSupportGitVersion)
+            {
+                throw new NotSupportedException($"MinSupported git version is {_minSupportGitVersion}, your request version is {version}");
+            }
+
+            if (!_gitCommands.ContainsKey(command))
+            {
+                throw new NotSupportedException($"Unsupported git command: {command}");
+            }
+
+            Dictionary<Version, String> options = _gitCommands[command];
+            foreach (var versionOption in options.OrderByDescending(o => o.Key))
+            {
+                if (version >= versionOption.Key)
+                {
+                    return versionOption.Value;
+                }
+            }
+
+            throw new NotSupportedException($"Can't find supported git command option for command: {command}, version: {version}.");
+        }
+
+        private async Task<int> ExecuteGitCommandAsync(IExecutionContext context, string repoRoot, string command, string options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            string arg = StringUtil.Format("{0} {1}", command, options ?? string.Empty).Trim();
```

Comment:
You don't need to coalesce with empty string. String.format doesn't null ref if an argument is null.